src/test/java/uk/gov/ch/oauth/Oauth2Test.java

```diff
@@ -191,6 +248,55 @@ public void testPrepareState() {
         assertEquals(STATE, state);
     }
 
+    @Test
+    @DisplayName("Test validate() returns true when correct User Profile is returned")
+    public void testValidateReturnsTrue() throws JsonProcessingException, InterruptedException {
+        when(oAuth2StateHandler.oauth2DecodeState(anyString()).toJSONObject().getAsString("nonce"))
+                .thenReturn(NONCE);
+
+        Map<String, Object> data = new HashMap<>();
+        data.put(SessionKeys.NONCE.getKey(), NONCE);
+        when(sessionFactory.getSessionDataFromContext()).thenReturn(data);
+
+        doReturn(oauthToken).when(oauth2).requestOAuthToken(anyString());
+        when(sessionFactory.regenerateSession()).thenReturn(session);
+        when(sessionFactory.buildSessionCookie(session)).thenReturn(cookie);
+        doReturn(userProfile).when(oauth2).requestUserProfile(oauthToken);
+        when(sessionFactory.getSessionFromContext()).thenReturn(session);
+
+        assertTrue(oauth2.validate(STATE, CODE, httpServletResponse));
+    }
+
+    @Test
+    @DisplayName("Test validate() returns false when fetchUserProfile() returns null due to an empty userProfile being returned")
+    public void testValidateReturnsFalse() throws JsonProcessingException, InterruptedException {
+        when(oAuth2StateHandler.oauth2DecodeState(anyString()).toJSONObject().getAsString("nonce"))
+                .thenReturn(NONCE);
+
+        Map<String, Object> data = new HashMap<>();
+        data.put(SessionKeys.NONCE.getKey(), NONCE);
+        when(sessionFactory.getSessionDataFromContext()).thenReturn(data);
+
+        doReturn(oauthToken).when(oauth2).requestOAuthToken(CODE);
+        when(sessionFactory.regenerateSession()).thenReturn(session);
+        when(sessionFactory.buildSessionCookie(session)).thenReturn(cookie);
+
+        UserProfileResponse userProfile = new UserProfileResponse();
+        doReturn(userProfile).when(oauth2).requestUserProfile(oauthToken);
+
+        assertFalse(oauth2.validate(STATE, CODE, httpServletResponse));
+    }
+
+    @Test
+    @DisplayName("Test validate() when getSessionNonce() throws a NullPointerException")
+    public void testGetSessionNonceThrowsNullPointer() {
+        when(oAuth2StateHandler.oauth2DecodeState(anyString()).toJSONObject().getAsString("nonce"))
+                .thenReturn(NONCE);
+        when(sessionFactory.getSessionDataFromContext()).thenReturn(data);
+        when(data.remove(SessionKeys.NONCE.getKey())).thenThrow(new NullPointerException());
```

Comment:
Is this the only place we've spied the data field for?

Reply:
Yeah the only place, would you like me to move it here locally?

Reply:
Why not just do it as real?

Reply:
As it won't necessary throw an exception in `getSessionNonce` which is what I want to test for.